enhancements/multi-arch/heterogeneous-architecture.md

```diff
@@ -0,0 +1,383 @@
+
+## Proposal
+
+The support for installation and upgrade of a heterogeneous architecture cluster and the associated components will be done in phases. The phases
+range from a bare bones no upgrade support installation to the full fledged installation and upgrade capability of a cluster on different topologies.
+The phases are listed below and details on some components of the phases are described below.
+
+Phase 0: Prototype
+- Aimed at delivering a prototype that users can experiment with.
+- Aimed at Azure/AWS with the primary supported architectures being x86 + ARM.
+- For supporting ARM nodes on Azure, [hyper-v support][BZ-1949614] is required which is landing in RHEL 8.6 and will be included in RHCOS for 4.11.
+- oc adm command changes to support manifestlist payloads.
+- Ability to assemble and publish manifestlist-based release payloads in a reproducible way.
+- Does not include upgrade support.
+- Is not intended for production.
+
+Phase 1: Tech Preview of basic support
+- Release pipeline needs to be setup to construct manifestlist based release payloads.
+- Upgrade support - CVO/Cincinnati support (heterogeneous -> heterogeneous payload upgrades only).
+- Does not include support for:
+  - image registry - internal image registry will not support manifest listed images.
+  - imagestreams importing full manifestlist content (will continue to import a single image when referencing manifestlists).
+
+Phase 2: Full basic functionality
+- Imagestream manifestlist support.
+- OLM support for node selection when deploying an operator and ecosystem improvements.
+- Improve console's content filtering by making it architecture aware and passing that information to OLM.
+- oc debug (and other commands) support for compute architecture.
+- support toolbox.
+- support for Power and Z architectures.
+- Hypershift support for heterogeneous management+worker clusters.
+
+Phase 3: Quality of life improvements
+- First class install support - ability to create non-control-plane-architecture workers at install time and generate machineconfig
+and machineset manifests for the same.
+- Possibly reduce mirroring costs by looking into things like sparse manifestlist.
+- Upgrade support for homogeneous payload -> heterogeneous release payload and vice versa.
+
+### User Stories
+
+- As an enterprise customer looking to save compute resource cost, I want to be able to run part of my application workload on ARM instead of x86.
+- As the owner of an application that runs partially on x86 and partially on ARM, I want to be able to run the entire application within a
+single cluster so I do not need to pay the cost of higher latency between clusters, or worry about the HA-resilience of cross-cluster bridges.
+
+### API Extensions
+
+- Imagestreams API will introduce a new field to indicate that it will reference/import manifestlists. It would also contain a new child manifest
+field which would be for the image corresponding to the control plane architecture.
```

Comment:
Can you include a golang like example of the changes you intend to make so that the API changes can be reviewed
eg
```
type ImageStreamsSpec struct {
  // Existing fields omitted for brevity

  // NewField is used to do, when it's not provided something happens, these are the boundaries and limitations of using the field
  // +kubebuilder:validation:...
  NewField []string `json:"newField"
```

Would be good to do this for both the changes mentioned in this section

Reply:
i think this is still in the works but @dmage should be able to provide something for the imagestreams and @wking can probably provide something for the ClusterVersion spec once they decide on the approach.

Reply:
i just came across this: https://github.com/openshift/enhancements/blob/master/enhancements/manifestlist/manifestlist-support.md#proposed-image-api-changes and i'm guessing the changes needed would be very similar.